Fix some issues with `powershell -Command -`

lzybkr opened the pull request:
Fixes #2071 and #2319.

lzybkr committed:
Fix `powershell -Command -` to not ignore last command

If the last command sent via stdin did not have a newline,
the command was ignored. This would happen when stdin was closed.

lzybkr committed:
Fix `powershell -Command -` handling of backspace

PowerShell can be started with input redirected in different scenarios.

`powershell -Command -` is one scenario where commands are sent to
PowerShell, but this is not a interactive shell scenario, it's really
a server to run commands from some other process.

`powershell -File -` or just regular redirected input means PowerShell
is an interactive shell, but command line editing must be done via
the redirected standard input handle instead of interacting directly
with a console handle.

In this scenario, we want to provide a good editing experience. Today,
we provide the bare minimum, supporting backspace but no cursor movement.

The bug here is that backspace was treated as a backspace in the server mode
but should not have been.

msftclas commented:
Hi __@lzybkr__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Jason Shirk (POWERSHELL)). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

lzybkr committed:
Fix hang when reading stdin from ruby

This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
They use named pipes instead of anonymous pipes, and for some reason that triggers the hang.

vors commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2868,6 +2868,25 @@ private class ConsoleHostStartupException : Exception
+            // This is a workaround for a full clr issue that causes a hang when calling PowerShell from ruby.
+            // They use named pipes instead of anonymous pipes, and for some reason that triggers a hang
+            // reading from Console.In.
+            var inputHandle = ConsoleControl.NativeMethods.GetStdHandle(-10);
```
What this `-10` mean? Can you add comment about it?

lzybkr replied:
Is it not obvious that it's standard in? (std handle, assigned to var named inputHandle, and comment mentions Console.In)

vors commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:
```diff
@@ -1588,7 +1591,9 @@ private string ReadLineFromFile(string initialContent)
-                if (c == '\b')
+                // If NoPrompt is true, we are in a sort of server mode where we shouldn't
+                // do anything like edit the command line - every character is part of the input.
+                if (c == '\b' && !NoPrompt)
```
Not sure about that. IO redirection scenario could be used to marshal interactive session as well, even with NoPrompt. I.e. in something like EditorServices. Did this caused mentioned bugs or it's just for tests?

PaulHigin reviewed (approved):
LGTM

ARBFL commented:
When is this fix available for Windows 1607? Will there be KB / fix in Windows Updates?

iSazonov commented:
@ARBFL This PR is closed.  You would be better to create Issue-Question.
